transformer_engine/pytorch/attention.py

```diff
@@ -481,91 +491,166 @@ def forward(ctx, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k
-                                    q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1],
-                                    cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
-                                    dropout_p, softmax_scale, causal=True, return_softmax=False,
+                            if use_fused_attention:
+                                # [b, 2, sq//2, np, hn] -> [b, sq, np, hn]
+                                q_inputs[i%2] = q.view(q.shape[0], -1, *q.shape[-2:])
+                                # [2, b, 2, sk//2, np, hn] -> [2, b, sk, np, hn]
+                                kv_inputs[i%2] = kv_inputs[i%2].view(2, k.shape[0], -1, *k.shape[-2:])
+                                out_per_step[i], [softmax_lse_per_step[i], rng_states[i]] = fused_attn_fwd(
+                                    is_training, max_seqlen_q, max_seqlen_k, cu_seqlens_q, cu_seqlens_k,
+                                    q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1], TE_DType[q.dtype],
+                                    tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen,
+                                    attn_scale=softmax_scale, dropout=dropout_p,
+                                    qkv_layout="bshd_bshd_bshd", attn_mask_type="causal",
                                 )
                             else:
-                                out_per_step[i] = torch.empty_like(q_inputs[i%2])
-                                _, softmax_lse_per_step[i], rng_states[i], _ = _flash_attn_forward( # pylint: disable=unbalanced-tuple-unpacking
-                                    q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1],
-                                    out_per_step[i], cu_seqlens_q, cu_seqlens_k,
-                                    max_seqlen_q, max_seqlen_k, dropout_p, softmax_scale,
-                                    causal=True, return_softmax=False,
-                                )
+                                # [b, 2, sq//2, np, hn] -> [b*sq, np, hn]
+                                q_inputs[i%2] = q.view(-1, *q.shape[-2:])
+                                # [2, b, 2, sk//2, np, hn] -> [2, b*sk, np, hn]
+                                kv_inputs[i%2] = kv_inputs[i%2].view(2, -1, *k.shape[-2:])
+                                if _flash_attn_2_available:
+                                    _, _, _, _, out_per_step[i], \
+                                    softmax_lse_per_step[i], _, rng_states[i] = _flash_attn_forward(
+                                        q_inputs[i%2], kv_inputs[i%2][0], kv_inputs[i%2][1],
+                                        cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
+                                        dropout_p, softmax_scale, causal=True, return_softmax=False,
+                                        **fa_optional_forward_kwargs
+                                    )
+                                else:
```

Comment:
We're dropping FAv1 support in PR 585, so `else` branches like this one might need to be deleted later. Just a headsup for you.